### ISSUE TITLE
Exception in @BeforeMethod || @AfterMethod declares method as Test case

### ISSUE THREAD

**Issue by echoAlexey** [allure-framework/allure-core#517]
Screenshots: 
http://prntscr.com/6fdcje
http://prntscr.com/6fdclu

Taken from TestNg example repository.

**Comment by lanwen** [comment]
And what is expected by you?

**Comment by echoAlexey** [comment]
We expected not to have setup method in Test case list. 

Now we see that exception should be shown somewhere.

We suggest show it in artificial test step. If that is difficult this issue can be closed.

**Comment by lanwen** [comment]
There are lot of problems how to show this methods - as they not part of any test case and executed only once. Now we generate artificial test case before any other - why artificial test step is better?

**Comment by volkovs** [comment]
Yes, agreed.

There are methods executed only once per test class. Other methods are executed once per test case. In later case we consider setup method as part of the test (that is why artificial test step might be better). But may be it doesn't worth considering both types of methods (complicating code base). Anyway, if setup method fails, we don't know the result of the test.

@lanwen what do you think about changing artificial test case status from CANCELED to BROKEN or FAILED (real test case is canceled, that's true) ?

**Comment by lanwen** [comment]
@volkovs bc-ac status is set by exception thrown. If you throw Assumption Exception - then it will be `CANCELED`. Think nothing should be changed in this logic.

**Comment by baev** [comment]
I propose to use `IInvokedMethodListener` and **always add configuration methods as separate test cases to the report**. That way we can handle steps and attachments in configuration methods. Also as I know these methods have status (which depends on throwing exception) and if some configuration methods fails other methods (and tests) get *canceled* status

**Comment by baev** [comment]
> always add configuration methods as separate test cases to the report

But `AfterMethod` and `BeforeMethod` run for each test case so I guess we need to merge it into test cases?

**Comment by kobrien** [comment]
I am also seeing this issue where the setup method is being counted as a "cancelled" test case when it fails.
I have annotated the setup/teardown methods with @BeforeClass and @AfterClass. 
When they pass successfully they do not show up in the report, and are not counted into the total test count (which is what I want)
It would be really great if we could show the setup/teardown in the report if its fails, but not to be counted into the total test case count.

**Comment by baev** [comment]
@kobrien sorry for the delay

> It would be really great if we could show the setup/teardown in the report if its fails, but not to be counted into the total test case count.

In general we doesn't know how much test cases are skipped when before class fixture is failed. And your test count can be different in different test runs. So we can doesn't count fake test cases but total count can be different anyway.

@lanwen @volkovs @allure-framework/qatools-team what guys are you think?

**Comment by volkovs** [comment]
@baev I think treating most configuration methods as test steps (instead of creating artificial test cases) would solve this and numerous other related issues we met (the only issue in that case I see so far is potential multiple failures in test case - e.g. both test and tear down methods). 

I will experiment with implementation a couple of days and let you know how it went. We have to implement it for ourselves anyway.

**Comment by baev** [comment]
@volkovs nice!

**Comment by volkovs** [comment]
@baev I have experimented with artificial Test Steps and have three points:

1. Please look at the reports generated:
![image](https://cloud.githubusercontent.com/assets/743546/8084800/abe4d64e-0f95-11e5-9c41-794c6fbba31f.png)
![image](https://cloud.githubusercontent.com/assets/743546/8084821/bdcd23ac-0f95-11e5-9009-e656b5e21be4.png)
![image](https://cloud.githubusercontent.com/assets/743546/8084834/cfe65324-0f95-11e5-9791-244eb44a702f.png)

As you can see `@BeforeTest` and `@BeforeClass` steps are injected in the first test case (first in suite). `@BeforeMethod` and `@AfterMethod` steps are injected in every test case. `@BeforeSuite` and `@AfterSuite` are injected as artificial test cases.

`@BeforeSuite` method broken or failed (status will change accordingly):
![image](https://cloud.githubusercontent.com/assets/743546/8084872/060d8918-0f96-11e5-8d6b-d7fe632ee592.png)

`@AfterClass` method broken or failed:
![image](https://cloud.githubusercontent.com/assets/743546/8084939/519a1e46-0f96-11e5-9236-6789a60c38b5.png)

`@AfterSuite` method broken or failed:
![image](https://cloud.githubusercontent.com/assets/743546/8084983/8e5b2d70-0f96-11e5-826d-de3a45e487e2.png)

2) Passed or Skipped configuration methods (configuration tests and test steps) could be hidden (NOT generated) with single configuration property (probably should be hidden by default).

3) `@BeforeSuite`/`@AfterSuite` (failure) will still create extra test thus this issue is not solved. We had another issue related - please look at report:
![image](https://cloud.githubusercontent.com/assets/743546/8085017/bdb1e2c6-0f96-11e5-8395-dbfb020192e2.png)

Since suite configuration method is not linked to any test it automatically appears in Behavior tab under Without feature -> Without story. That is confusing for QA engineers who track unmapped tests.
That is why I start thinking to treat suite configuration methods in the same way as any other one. That will solve both issues and simplify the code. What do you think?

P.S. It may also solve #419. But that is a separate topic to discuss.

**Comment by baev** [comment]
I personally like to add to `TestCaseResult` information about fixtures. An example:

```xsd
<xs:complexType name="test-case-result">
    <xs:all>
        <xs:element name="name" type="xs:string"/>
        <xs:element name="title" type="xs:string" minOccurs="0"/>
        <xs:element name="description" type="qa:description" minOccurs="0"/>
        <xs:element name="failure" type="qa:failure" minOccurs="0"/>
        
        <xs:element name="before-fixtures" type="qa:steps" minOccurs="0"/>
        <xs:element name="steps" type="qa:steps" minOccurs="0"/>
        <xs:element name="attachments" type="qa:attachments" minOccurs="0"/>
        <xs:element name="after-fixtures" type="qa:steps" minOccurs="0"/>
        
        <xs:element name="labels" type="qa:labels" minOccurs="0"/>
        <xs:element name="parameters" type="qa:parameters" minOccurs="0"/>
    </xs:all>
    <xs:attribute name="start" type="xs:long" use="required"/>
    <xs:attribute name="stop" type="xs:long" use="required"/>
    <xs:attribute name="status" type="qa:status" use="required"/>
    <xs:attribute name="severity" type="qa:severity-level">
        <xs:annotation>
            <xs:documentation>This is Deprecated. Use label with name 'severity' instead.</xs:documentation>
        </xs:annotation>
    </xs:attribute>
</xs:complexType>
```

**Comment by volkovs** [comment]
Yes, I like this idea as well. That would help visually separate configuration steps from test steps. Test events generation wouldn't be much different.

**Comment by baev** [comment]
@volkovs but I am a bit confused about test group fixtures (such as `@BeforeClass`, `@BeforeSuite`, and '@BeforeGroups'). Probably we should copy such fixtures to each test case..

**Comment by volkovs** [comment]
@baev copying fixture will make impression they were executed multiple times for every test case. In that case there is extra model and UI complexities to separate "executed steps" (like `@BeforeMethod`) from "information or context steps" (steps we are currently in scope of). I mean users would like to know exactly how many times each configuration was executed.

In our adapter we simply agreed to display steps in configuration execution order in the "nearest" test case (as in screenshots above). I can update you on the users feedback once we have it.

**Comment by volkovs** [comment]
I am excited about fixtures in the model. In that case there would be new FixtureStartedEvent instead of StepStartedEvent. So we could transparently add support of fixture test steps before TestCaseStartedEvent and after TestCaseFinishedEvent (currently we use workaround - updating model directly, since test step events don't work out of test case scope). What do you think?

**Comment by braimanm** [comment]
PR #598  "Fixing #569 and other testNg cumulative issues" will fix this issue. The failing or cancelled fixture methods will appear in the report as independent test cases.

**Comment by baev** [comment]
closed due to no activity

**Comment by velzganesh** [comment]
@baev Why is this ticket closed? I am facing this issue every single day.

**Comment by baev** [comment]
@velzganesh hi! I don't think we gonna fix that issue in 1.4. So I like to keep it closed. BTW it was already added to roadmap for Allure 2. So stay turned!

**Comment by aTan-aka-Xellos** [comment]
@baev where we can see the road-map? And when approximately 2.0 is planned? 
Our project face the same issue and to have correct numbers is highly important...
